streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java

```diff
@@ -0,0 +1,148 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals.graph;
+
+import org.apache.kafka.streams.kstream.internals.ConsumedInternal;
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder;
+import org.apache.kafka.streams.state.StoreBuilder;
+
+import java.util.Collections;
+
+/**
+ * Used to represent either a KTable source or a GlobalKTable source. A boolean flag is used to indicate if this represents a GlobalKTable a {@link
+ * org.apache.kafka.streams.kstream.GlobalKTable}
+ */
+public class TableSourceNode<K, V> extends StreamSourceNode<K, V> {
+
+    private StoreBuilder storeBuilder;
+    private final ProcessorParameters<K, V> processorParameters;
+    private final String sourceName;
+    private final boolean isGlobalKTable;
+
+    TableSourceNode(final String nodeName,
+                    final String sourceName,
+                    final String topic,
+                    final ConsumedInternal<K, V> consumedInternal,
+                    final StoreBuilder storeBuilder,
+                    final ProcessorParameters<K, V> processorParameters,
+                    final boolean isGlobalKTable) {
+
+        super(nodeName,
+              Collections.singletonList(topic),
+              consumedInternal);
+
+        this.processorParameters = processorParameters;
+        this.sourceName = sourceName;
+        this.isGlobalKTable = isGlobalKTable;
+        this.storeBuilder = storeBuilder;
+    }
+
+    public static <K, V> TableSourceNodeBuilder<K, V> tableSourceNodeBuilder() {
+        return new TableSourceNodeBuilder<>();
+    }
+
+    @Override
+    public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
+        String topicName = getTopicNames().get(0);
+
+        if (isGlobalKTable) {
+            topologyBuilder.addGlobalStore(storeBuilder,
+                                           sourceName,
+                                           consumedInternal().timestampExtractor(),
+                                           consumedInternal().keyDeserializer(),
+                                           consumedInternal().valueDeserializer(),
+                                           topicName,
+                                           processorParameters.processorName(),
+                                           processorParameters.processorSupplier());
+        } else {
+            topologyBuilder.addSource(consumedInternal().offsetResetPolicy(),
+                                      sourceName,
+                                      consumedInternal().timestampExtractor(),
+                                      consumedInternal().keyDeserializer(),
+                                      consumedInternal().valueDeserializer(),
+                                      topicName);
+
+            topologyBuilder.addProcessor(processorParameters.processorName(), processorParameters.processorSupplier(), sourceName);
+
```

Comment:
For now, this includes the recent changes from @guozhangwang (PR #5163) for optional re-use of source topic as changelog topic.  This optimization will get folded into this PR in a follow-up push.

Reply:
I cannot tell how this includes the source topic reuse logic, could you explain a bit?